modules/ctrl_tcp/ctrl_tcp.c

```diff
@@ -0,0 +1,358 @@
+/**
+ * @file ctrl_tcp.c  TCP control interface using JSON payload
+ *
+ * Copyright (C) 2017 José Luis Millán <jmillan@aliax.net>
+ */
+#include <re.h>
+#include <baresip.h>
+
+#include "netstring.h"
+
+
+/**
+ * @defgroup ctrl_tcp ctrl_tcp
+ *
+ * Communication channel to control and monitor Baresip via JSON messages.
+ *
+ * It receives commands to be executed, sends back command responses and
+ * notifies about events.
+ *
+ * Command message parameters:
+ *
+ * - command : Command to be executed.
+ * - params  : Command parameters.
+ * - token   : Optional. Included in the response if present.
+ *
+ * Command message example:
+ *
+ \verbatim
+ {
+  "command" : "dial",
+  "params"  : "sip:alice@atlanta.com",
+  "token"   : "qwerasdf"
+ }
+ \endverbatim
+ *
+ *
+ * Response message parameters:
+ *
+ * - type     : "response". Identifies the message type.
+ * - response : Baresip response to the related command execution.
+ * - token    : Present if it was included in the related command request.
+ *
+ * Response message example:
+ *
+ \verbatim
+ {
+  "type"     : "response",
+  "response" : "",
+  "token"    : "qwerasdf"
+ }
+ \endverbatim
+ *
+ *
+ * Event message parameters:
+ *
+ * - type      : "event". Identifies the message type.
+ * - class     : Event class.
+ * - event     : Event ID.
+ * - param     : Specific event information.
+ *
+ * Apart from the above, events may contain aditional parameters.
+ *
+ * Event message example:
+ *
+ \verbatim
+ {
+  "type"       : "event",
+  "class"      : "call",
+  "event"      : "CALL_CLOSED",
+  "param"      : "Connection reset by peer",
+  "account"    : "sip:alice@atlanta.com",
+  "direction"  : "incoming",
+  "peer"       : "sip:bob@biloxy.com",
+  "id"         : "73a12546589651f8"
+ }
+ \endverbatim
+ *
+ *
+ * Sample config:
+ *
+ \verbatim
+  ctrl_tcp_listen     0.0.0.0:4444         # IP-address and port to listen on
+ \endverbatim
+ */
+
+
+enum {CTRL_PORT = 4444};
+
+struct ctrl_st {
+	struct tcp_sock *ts;
+	struct tcp_conn *tc;
+	struct netstring *ns;
+};
+
+static struct ctrl_st *ctrl = NULL;  /* allow only one instance */
+
+static int print_handler(const char *p, size_t size, void *arg)
+{
+	struct mbuf *mb = arg;
+
+	return mbuf_write_mem(mb, (uint8_t *)p, size);
+}
+
+
+static int encode_response(struct mbuf *resp, const char *token)
+{
+	struct re_printf pf = {print_handler, resp};
+	struct odict *od = NULL;
+	char *buf = NULL;
+	int err;
+
+	err = odict_alloc(&od, 8);
+	if (err)
+		return err;
+
+	resp->pos = NETSRING_HEADER_SIZE;
```

Comment:
is there a typo in `NETSRING_HEADER_SIZE` ?

Reply:
Indeed :-). I'll fix it.